### DISCUSSION TITLE
rust-course/advance/macro

### DISCUSSION THREAD

**Discussion by sunface** [sunface/rust-course#631] [章节评论区]
https://course.rs/advance/macro.html

**Comment by gzeronet** [comment]
通俗易懂，感谢。

**Comment by dafu-wu** [comment]
赞，意犹未尽

**Comment by qq906907952** [comment]
过程宏的描述和例子太简单了，这里面还有很多内容，例如怎么去解析语法树。

**Reply by starrys147** [reply]
别的地方学吧，一般情况下除非开发一个crate否则用不到这些知识的（主要是开发一个过程宏太麻烦了，而且在项目中也不一定用得到。）
而要是你都能开发一个包了，还看这个入门教程干嘛（）

**Reply by zy010101** [reply]
如无必要，不要开发一个过程宏。开发和维护过程宏很麻烦。

**Comment by lnb-wxqy** [comment]
一脸懵逼

**Comment by csx12588** [comment]
建议把最后的一段话放在最前面

**Comment by iscas-zac** [comment]
```rust
pub fn my_default(input: TokenStream) -> TokenStream {
     ...
}
```

里面那段代码完整吗？<code>field_ast</code> 在返回值里没有用到，for 循环中的两个 <code>quote!</code> 也是空的。如果不完整能不能加个注解

**Reply by qq906907952** [reply]
这个应该是md解析的问题，# 后面的消失了 不知道什么原因

**Comment by wyb10a10** [comment]
示例代码有错
```rust
use proc_macro::TokenStream;
use syn::{self, Data};
use quote::{quote};
use syn::{parse_macro_input, DeriveInput};

#[proc_macro_derive(MyDefault)]
pub fn my_default(input: TokenStream) -> TokenStream {
    let ast: DeriveInput = syn::parse(input).unwrap();
    let id = ast.ident;

    if let Data::Struct(s) = ast.data {
        // 声明一个新的ast，用于动态构建字段赋值的token
        let mut field_ast = quote! {};

        // 这里就是要动态添加token的地方了，需要动态完成Self的字段赋值
        for (idx, f) in s.fields.iter().enumerate() {
            let (field_id, field_ty) = (&f.ident, &f.ty);

            if let Some(field_id) = field_id {
                // 对于命名字段，都需要添加 `#field_name: #field_type::default(),` 这样的代码
                field_ast.extend(quote! { #field_id: <#field_ty>::default(), });
            } else {
                let field_idx = syn::Index::from(idx);
                let fty = &f.ty;
                //对于匿名字段，都需要添加 `#field_idx: <#field_type as std::default::Default>::default(),` 这样的代码
                field_ast.extend(quote! { #field_idx: <#fty>::default(), });
            }
        }

        let result = quote! {
            impl std::default::Default for #id {
                fn default() -> Self {
                    Self { #field_ast }
                }
            }
        }
        .into();

        result
    } else {
        panic!("MyDefault derive macro must use in struct");
    }
}
```

**Reply by qq906907952** [reply]
你是指少了 
#field_id: <#field_ty>::default()
#field_idx: <#fty>::default()
#field_ast
这三行？ 这个源码是有的，不知道为什么这里 # 后面的消失了

**Reply by Pro741** [reply]
是因为那个小眼睛!, 它把所有以`#`开头的这一行代码给隐藏了

**Comment by yuhao00** [comment]
```rust
($($x:expr),*)
```
这种方式并不能匹配结尾逗号

**Comment by roushanyyzz** [comment]
少了点代码
```rust
#[proc_macro_derive(MyDefault)]
pub fn my_default(input: TokenStream) -> TokenStream {
    let ast: DeriveInput = syn::parse(input).unwrap();
    let id = ast.ident;

    let Data::Struct(s) = ast.data else {
        panic!("MyDefault derive macro must use in struct");
    };

    let mut field_ast = quote!();

    for (idx, f) in s.fields.iter().enumerate() {
        let (field_id, field_ty) = (&f.ident, &f.ty);

        // 逗号不要忽略了
        if field_id.is_none() {
            let field_idx = syn::Index::from(idx);
            field_ast.extend(quote! {#field_idx:<#field_ty>::default(),})
        } else {
            field_ast.extend(quote! {#field_id:<#field_ty>::default(),})
        }
    }

    quote! {
        impl Default for #id{
            fn default() -> Self{
                Self {#field_ast}
            }
        }
    }
    .into()
}
```

**Comment by Xiaobaishushu25** [comment]
宏强大归强大，但是用起来ide没有提示也太草蛋了，跟用记事本写一样，clion和vscode都不太好使

**Comment by uyplayer** [comment]
```
 ( $( $x:expr ),* )
```
上面的宏无法匹配
```
1, 2, 3,
```
只匹配一下的表达式
```
1, 2, 3
```

如果想匹配
```
1, 2, 3,
```
那就表达式
```
($($x:expr,)*)
```

**Reply by Pro741** [reply]
一般的解决方法是 ( $( $x:expr ),* $(,)? )

**Comment by rucat125** [comment]
好了，成功地不认识“宏”字了

**Comment by tjw123hh** [comment]
《`hell_macro()`》

**Comment by ptrzs3** [comment]
`cargo expand`命令发生变化了，应该执行`cargo expand --bin hello_macro`，具体用法`cargo expand --help`

**Reply by ZLD1819858** [reply]
thanks.

**Comment by Ray-lyy** [comment]
好难，不求甚解过一遍都好难，以后如果用到了再回来翻可能会好点

**Comment by ceamanz** [comment]
提点小建议：

模式解析
第3点：$() 之后的逗号，意味着1 和 2 之间可以使用逗号进行分割，也意味着 3 既可以没有逗号，也可以有逗号：vec![1, 2, 3,]

上面这段话有点小问题，按本文例子，3后面如果有括号是匹配不到的，会报错；只有3后面没有逗号才可以通过编译。
实际应该写成：( $( $k:expr ),* $(,)* ) 才能匹配到最后一个元素后面的逗号。

感谢此文，声明宏终于入门了一点点，功德无量。

**Comment by iamtaojin** [comment]
通俗，难懂！

**Comment by pk111and222** [comment]
居然如此强大，岂不是可以利用宏的结构来自己定义一套语言了

**Reply by Wyatex** [reply]
感觉是的，只要会写parser就啥语言都能转rust😂

**Comment by LoserJL** [comment]
把宏设计成这样，真是个鬼才，设计的人不经常使用岂不是浪费了

**Comment by QuarkPixel** [comment]
> 这里就比较好理解了，$() 中的 temp_vec.push() 将根据模式匹配的次数生成对应的代码，当调用 vec![1, 2, 3] 时，下面这段生成的代码将替代传入的源代码，也就是替代 vec![1, 2, 3] :
> ```
> {
>     let mut temp_vec = Vec::new();
>     temp_vec.push(1);
>     temp_vec.push(2);
>     temp_vec.push(3);
>     temp_vec
> }
```
这里为什么是一个一个 `push()` 的？这样是否会造成由于 capacity 不够导致的反复开辟新空间从而造成的性能浪费？为什么这里没有用 `Vec::new(with_capacity)` 呢？还是说这段代码在编译阶段就会被优化之类的？

**Reply by QuarkPixel** [reply]
写错了，刚才写的时候没检查，应该是
`Vec::with_capacity(n);`

**Reply by QuarkPixel** [reply]
刚才对比下源代码又回来看，发现这个宏只是一个简单实现，算是我没有认真看了😂

**Comment by zhang12295** [comment]
error: missing parameters for function definition
 --> src/main.rs:4:20
  |
4 | #[derive(HelloMacro)]
  |                    ^
  |
  = note: this error originates in the derive macro `HelloMacro` (in Nightly builds, run with -Z macro-backtrace for more info)

依赖导入不了怎么解决，求大佬指点

**Comment by WowYouCanReallyDance** [comment]
作者大大，有个地方我觉得可以明晰一下，就是cargo的最小编译单元是crate（箱），过程宏必须至少放在一个library crate中。
包也就是package是crate的上级组织结构。
其实proc-macro是可以和普通代码共存于同一个package中的，只是proc-macro代码必须放在单独的library crate中。

**Comment by reatang** [comment]
这一篇搞晕我了，我来重新给宏分类：

传统宏：是C语言那种老式的宏概念
过程宏：基于现代AST语法树的宏概念

1️⃣ 按定义方式
传统宏：macro_rules! name { ... } → 定义函数宏的规则
过程宏：
#[proc_macro] fn name(...) → 自定义函数宏
#[proc_macro_attribute] fn name(attr, item) → 属性宏
#[proc_macro_derive] fn name(...) → derive 宏

这个维度告诉你“怎么创建宏”，是动作层面的分类。

2️⃣ 按调用方式
函数宏：name!(...) → 调用生成代码
属性宏：#[...] → 标注在 item 上
内层属性：#![...] → 放在 crate 或模块开头

这个维度告诉你“在代码里怎么用宏”，是使用层面的分类。

3️⃣ 按作用 / 作用域
宏类型	作用对象	输出	示例
属性宏	struct / fn / enum / module	修改或生成整个 item	#[tokio::main]
派生宏	struct / enum	生成 trait impl	#[derive(Clone, Debug)]
函数宏	表达式 / 语句	生成代码片段	println!("Hi")、vec![1,2,3]
内层属性	crate / module	配置 / 指令	#![allow(dead_code)]

这个维度告诉你“宏能影响哪些东西”，是作用层面的分类。

**Comment by CaoChunfang** [comment]
这里的过程宏是不是相当于Java中使用一个注解进行标记，然后再通过AOP的方式进行增强？

**Reply by TwinklerG** [reply]
效果类似，但实现完全不同。Rust过程宏是编译期完成，而Java注解中，IOC通过反射完成、AOP通过Proxy完成，均在运行时进行。

**Comment by zhao-xiaogang** [comment]
**受益匪浅**。

随 rust 版本更新，过程宏 `HelloMacro` 的例子需要修改一下才能跑通：
### 1. gen 是新版本的保留字
```
error: expected identifier, found reserved keyword `gen`
  --> hello_macro_derive/src/lib.rs:19:6
   |
19 |     let gen = quote! {
   |         ^^^ expected identifier, found reserved keyword
   |
```
修改 hello_macro_derive/Cargo.toml 中版本为 2021 可以解决。
```
edition = "2021"
```

### 2. HelloMacro 需要显式导入
```
error: cannot find derive macro `HelloMacro` in this scope
 --> src/main.rs:4:10
  |
4 | #[derive(HelloMacro)]
  |          ^^^^^^^^^^
  |
note: `HelloMacro` is imported here, but it is only a trait, without a derive macro
 --> src/main.rs:2:5
```
错误原因：
  在 src/main.rs 中，`use hello_macro::HelloMacro`; 只导入了 trait，但 #[derive(HelloMacro)] 需要的是 derive
  macro。trait 和 derive macro 虽然同名，但属于不同的命名空间。
  修复方法：
  在 hello_macro/src/lib.rs:5 添加了一行：`pub use hello_macro_derive::HelloMacro;`
  这样 hello_macro crate 在导出 trait 的同时，也重新导出了 derive macro。main.rs 里的 use
  hello_macro::HelloMacro; 就能同时引入两者，#[derive(HelloMacro)] 就可以正常工作了。

**Reply by zhao-xiaogang** [reply]
第 2 点请忽略，是我自己写漏了。
